spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java

```diff
@@ -0,0 +1,377 @@
+    Assert.assertEquals((short) -2, scalarSql("SELECT system.truncate(2, -1S)"));
+
+    // Check that short types are allowed for the width
+    Assert.assertEquals((short) 0, scalarSql("SELECT system.truncate(5S, 1S)"));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS smallint))"));
+  }
+
+  @Test
+  public void testTruncateInt() {
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 0));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 1));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 5));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 9));
+    Assert.assertEquals(10, scalarSql("SELECT system.truncate(10, %d)", 10));
+    Assert.assertEquals(10, scalarSql("SELECT system.truncate(10, %d)", 11));
+    Assert.assertEquals(-10, scalarSql("SELECT system.truncate(10, %d)", -1));
+    Assert.assertEquals(-10, scalarSql("SELECT system.truncate(10, %d)", -5));
+    Assert.assertEquals(-10, scalarSql("SELECT system.truncate(10, %d)", -10));
+    Assert.assertEquals(-20, scalarSql("SELECT system.truncate(10, %d)", -11));
+
+    // Check that different widths can be used
+    Assert.assertEquals(-2, scalarSql("SELECT system.truncate(2, %d)", -1));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(300, %d)", 1));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS int))"));
+  }
+
+  @Test
+  public void testTruncateBigInt() {
+    Assert.assertEquals(0L, scalarSql("SELECT system.truncate(10, 0L)"));
+    Assert.assertEquals(0L, scalarSql("SELECT system.truncate(10, 1L)"));
+    Assert.assertEquals(0L, scalarSql("SELECT system.truncate(10, 5L)"));
+    Assert.assertEquals(0L, scalarSql("SELECT system.truncate(10, 9L)"));
+    Assert.assertEquals(10L, scalarSql("SELECT system.truncate(10, 10L)"));
+    Assert.assertEquals(10L, scalarSql("SELECT system.truncate(10, 11L)"));
+    Assert.assertEquals(-10L, scalarSql("SELECT system.truncate(10, -1L)"));
+    Assert.assertEquals(-10L, scalarSql("SELECT system.truncate(10, -5L)"));
+    Assert.assertEquals(-10L, scalarSql("SELECT system.truncate(10, -10L)"));
+    Assert.assertEquals(-20L, scalarSql("SELECT system.truncate(10, -11L)"));
+
+    // Check that different widths can be used
+    Assert.assertEquals(-2L, scalarSql("SELECT system.truncate(2, -1L)"));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS bigint))"));
+  }
+
+  @Test
+  public void testTruncateDecimal() {
+    // decimal truncation works by applying the decimal scale to the width: ie 10 scale 2 = 0.10
+    Assert.assertEquals(
+        new BigDecimal("12.30"),
+        scalarSql("SELECT system.truncate(10, CAST(%f as DECIMAL(9, 2)))", 12.34));
```

Comment:
When working with decimals, you want to prefer strings. Formatting a decimal may give a different value and it looks like this assumes the value will be what was given as a float.

Reply:
Good point. Updated to use format strings with string values for the decimal test cases.